refactor: implement DI in team billing service and team billing data  repository factory

joeauyeung opened the pull request:
## What does this PR do?

This PR refactors the team/organization billing system to use dependency injection (DI) with the `ioctopus` library, following SOLID principles. The main changes include:

- **Implements DI containers** for billing services using `ioctopus`
- **Renames files** from kebab-case to camelCase for consistency across the codebase
- **Creates factory pattern** for `TeamBillingService` and `BillingProviderService` that automatically returns stub implementations when billing is disabled
- **Consolidates billing logic** into better organized service and repository layers
- **Updates all consumers** (60 files) to use factory functions instead of direct instantiation

**Key architectural improvements:**
- `getTeamBillingServiceFactory()` - Factory for team billing operations
- `getBillingProviderService()` - Factory for Stripe billing service
- `getTeamBillingDataRepository()` - Repository for team billing data access
- Stub implementations automatically used when `IS_TEAM_BILLING_ENABLED=false`

## Mandatory Tasks (DO NOT REMOVE)

- [x] I have self-reviewed the code
- [x] I have updated the developer docs in /docs if this PR makes changes that would require a documentation change. **N/A** - Internal refactoring only
- [ ] I confirm automated tests are in place that prove my fix is effective or that my feature works. **Partial** - Updated existing test for `skipTeamTrials`, but this is primarily a refactoring PR

## How should this be tested?

**Environment Setup:**
- Ensure `IS_TEAM_BILLING_ENABLED` is set appropriately
- Ensure `STRIPE_PRIVATE_KEY` is configured
- New env var `STRIPE_ORG_PRODUCT_ID` added to turbo.json

**Test Scenarios:**

1. **Team Billing Operations** (with billing enabled):
   - Create a new team with a paid plan
   - Invite members to a team (should update quantity)
   - Delete users from an organization (should update quantity)
   - Verify subscription status checks work correctly

2. **Billing Disabled Mode**:
   - Set `IS_TEAM_BILLING_ENABLED=false`
   - Verify stub implementations are used (no actual Stripe calls)
   - Team operations should still work without errors

3. **Webhook Handling**:
   - Test `customer.subscription.deleted` webhook
   - Test `invoice.paid` webhook for organizations

4. **Subscription Status**:
   - Verify `hasActiveTeamPlan` correctly identifies active/trialing/past_due subscriptions
   - Test `skipTeamTrials` for teams in trial period

**Expected Behavior:**
- All existing billing functionality should work identically
- No changes to business logic, only architectural improvements
- Stub services should be transparent when billing is disabled

## Review Checklist

**Critical areas to review:**

1. **Factory Usage** - Verify all instances of `new TeamBilling()`, `new InternalTeamBilling()`, and `new StripeBillingService()` have been replaced with factory calls
2. **Subscription Status Enum** - Check that the change from `Stripe.Subscription.Status` (lowercase strings) to `SubscriptionStatus` enum (uppercase) is handled correctly in all comparisons
3. **Stub Implementation** - Verify stub services are returned when `IS_TEAM_BILLING_ENABLED=false` and don't cause runtime errors
4. **Repository Pattern** - Ensure the repository factory correctly switches between Prisma and stub implementations
5. **Webhook Handlers** - Verify webhook handlers still work with the new factory pattern
6. **Test Coverage** - Only one test file was updated; consider if more test updates are needed

**Known Risks:**
- Large refactor (60 files) increases risk of subtle bugs
- Billing is critical functionality - any issues could affect payments
- Type changes in subscription status handling need careful verification

---


**Link to Devin run:** https://app.devin.ai/sessions/35e0b25ba04b4f59b50b4fef1a492411  
**Requested by:** joe@cal.com (@joeauyeung)

joeauyeung committed:
Move TeamBillingRepositories

joeauyeung committed:
WIP refactor team internal billing service

joeauyeung committed:
Remove duplicate billing repository files

joeauyeung committed:
Remove logic check in repository for billing is enabled

joeauyeung committed:
Rename repository to `TeamBillingData`

joeauyeung committed:
Use repository factory in main service

joeauyeung committed:
Fix new import paths

joeauyeung commented on packages/features/ee/billing/repository/teamBillingData/teamBillingDataRepositoryFactory.ts:
New factory to return the prisma repository or the stub one if billing is not enabled

joeauyeung committed:
Merge branch 'main' into implement-team-billing-repository-factor

joeauyeung committed:
Rename  to

joeauyeung committed:
Ensure `IS_TEAM_BILLING_ENABLED` is of type boolean

joeauyeung committed:
Rename classes to TeamBillingService and TeamBillingServiceFactory

joeauyeung committed:
Implement DI in `BookingServiceFactory`

joeauyeung committed:
`TeamBillingService` use repository in `getOrgIfNeeded`

joeauyeung committed:
DI `isTeamBillingEnabled` to `TeamBillingServiceFactory`

joeauyeung committed:
Rename files for consistency

joeauyeung committed:
Return stub BillingRepository if billing is not enabled

joeauyeung committed:
Move Stripe billing service to service folder

joeauyeung committed:
Rename file

joeauyeung committed:
`StripeBillingService.getSubscriptionStatus` return `SubscriptionStatus`

joeauyeung committed:
Type fices in StripeBillingService

joeauyeung committed:
Type fix in `stubTeamBillingService`

joeauyeung committed:
DI the `BillingProviderService` into the `TeamBillingService`

joeauyeung committed:
Implement DI in `skipTeamTrials.handler`

joeauyeung committed:
Implement DI for team billing in `inviteMember.handler`

joeauyeung committed:
`skipTeamTrials.handler` use `team.isOrganization`

joeauyeung committed:
Implement DI for billing in `hasActiveTeamPlan.handler`

joeauyeung committed:
Type fixes

joeauyeung committed:
Implement DI in `bulkDeleteUsers.handler`

joeauyeung committed:
Implement `BillingProviderServiceFactory` in `updateProfile.handler`

joeauyeung committed:
Implment `BillingProviderServiceFactory` in `buyCredits.handler`

joeauyeung committed:
Fix import in `stripeCustomer.handler`

joeauyeung committed:
Add a constructor to `teamBillingServiceFactory`

joeauyeung committed:
Add DI to `PrismaTeamBillingRepository`

joeauyeung committed:
Add DI to `StripeBillingService`

joeauyeung committed:
Implement singleton in `BillingProviderServiceFactory`

joeauyeung committed:
Add DI folder and contents to billing folder

joeauyeung committed:
Use `getTeamBillingServiceFactory` in `inviteMember.handler`

joeauyeung committed:
Add `saveTeamBilling` method to `ITeamBillingService`

joeauyeung committed:
Implement DI in new team route

joeauyeung committed:
Implement DI in `teamService`

joeauyeung committed:
Implement DI in `OrganizationPaymentService`

joeauyeung committed:
Implement DI in `credit-service`

joeauyeung committed:
In `StripeBillingService` remove `static` from status methods

joeauyeung committed:
Implemnt DI in `_invoice.paid.org`

joeauyeung committed:
Refactor `hasActiveTeamPlan` to use `getTeamBillingFactory`

joeauyeung committed:
Refactor `skipTeamTrials` to use `getTeamBillingFactory`

joeauyeung committed:
Refactor `skipTeamTrials` to use `getTeamBillingServiceFactory`

joeauyeung committed:
`stripeCustomer.handler` to use `getBillingProviderService`

joeauyeung committed:
Remove old factories

joeauyeung committed:
Type fix

joeauyeung committed:
Remove unused factory

joeauyeung committed:
Refactor `updateProfile.handler` to use `getBillingProviderService`

joeauyeung committed:
Change name to `TeamBillingDataRepositoryFactory`

joeauyeung committed:
Type Prisma return in `prisma.module`

joeauyeung committed:
Type fix

joeauyeung committed:
Refactor `buyCredits.handler` to use `getBillingProviderService`

joeauyeung committed:
Refactor `credit-service` to use billing DI containers

joeauyeung committed:
Type fix

joeauyeung committed:
Add `getTeamBillingDataRepository`

joeauyeung committed:
Refactor `_invoice.paid.org` to use DI container

joeauyeung committed:
Refactor `_customer.subscription.deleted.team-plan` to use DI container

joeauyeung committed:
Refactor `calcomHandler` to use DI container

joeauyeung committed:
Refactor `getCustomerAndCheckoutSession` to use DI container

joeauyeung committed:
Refactor `verify-email` to use DI containers

joeauyeung committed:
Refactor `api/create/route` to use DI container

joeauyeung committed:
Refactor downgradeUsers to use DI container

joeauyeung committed:
Merge

joeauyeung committed:
Type fix

joeauyeung committed:
Clean up console.logs

joeauyeung committed:
Add await to `this.billingRepository.create` in `saveTeamBilling`

Co-authored-by: cubic-dev-ai[bot] <191113872+cubic-dev-ai[bot]@users.noreply.github.com>

joeauyeung committed:
Fix type errors

joeauyeung committed:
Address comments

devin-ai-integration[bot] committed:
fix: update tests to work with new DI pattern

- Update teamBillingService.test.ts to properly inject DI dependencies
- Remove unused billingModule import and mock
- Fix import naming in teamService.integration-test.ts (remove unused rename)
- Fix import path for TeamBillingPublishResponseStatus

All tests now properly mock IBillingProviderService, ITeamBillingDataRepository,
and IBillingRepository instead of using the old BillingRepositoryFactory pattern.

Co-Authored-By: joe@cal.com <j.auyeung419@gmail.com>

devin-ai-integration[bot] committed:
fix: add compatibility layer and env setup for unit tests

- Add STRIPE_PRIVATE_KEY dummy value to vitest.config.ts to prevent DI module errors
- Fix import paths in credit-service.test.ts (StripeBillingService, TeamBillingService)
- Create compatibility barrel at packages/features/ee/billing/teams/index.ts for test mocking

Co-Authored-By: joe@cal.com <j.auyeung419@gmail.com>

devin-ai-integration[bot] committed:
fix: update unit tests to mock DI container properly

- Update teamService.test.ts to mock getTeamBillingServiceFactory() instead of TeamBilling.findAndInit
- Update teamService.alternative.test.ts to mock DI container
- Update credit-service.test.ts to mock getBillingProviderService() and use SubscriptionStatus enum values
- Update OrganizationPaymentService.test.ts to mock DI container instead of direct StripeBillingService import
- Remove all 'as any' type casting to comply with Cal.com coding standards
- Fix unused variable warnings by prefixing with underscore

All 53 tests now passing (16 + 1 + 30 + 6)

Co-Authored-By: joe@cal.com <j.auyeung419@gmail.com>

joeauyeung committed:
Merge branch 'main' into implement-team-billing-repository-factor

devin-ai-integration[bot] committed:
fix: update remaining unit tests to use DI pattern

- Fix StripeBillingService.test.ts to inject mock Stripe client directly
- Fix teamBillingFactory.test.ts to mock getTeamBillingServiceFactory() from DI container
- Fix skipTeamTrials.test.ts to mock DI container and use SubscriptionStatus enum

All 11 previously failing tests now pass (5 + 5 + 1)

Co-Authored-By: joe@cal.com <j.auyeung419@gmail.com>

joeauyeung committed:
Undo changes made to Prisma module

joeauyeung commented on packages/features/ee/billing/di/modules/BillingRepositoryFactory.ts:
```diff
@@ -0,0 +1,43 @@
+  return (isOrganization: boolean): IBillingRepository => {
+    if (!isTeamBillingEnabled) {
+      return new StubBillingRepository();
+    }
```
Factory handles logic around returning a stub or an actual repository

joeauyeung commented on packages/features/ee/billing/service/teams/teamBillingService.ts:
```diff
@@ -11,23 +11,46 @@ import { prisma } from "@calcom/prisma";
-  constructor(team: TeamBillingInput) {
+  private teamBillingDataRepository: ITeamBillingDataRepository;
+
+  constructor({
```
Add a constructor here to pass the dependencies

devin-ai-integration[bot] committed:
fix: address test-related PR comments

- Fix OrganizationPaymentService.test.ts mock path from @calcom/ee to @calcom/features/ee
- Refactor teamBillingFactory.test.ts to test real factory logic instead of mocking container
- Remove duplicate teamBillingService.test..ts file with incorrect double-dot filename

All three test files now pass successfully with proper DI patterns.

Co-Authored-By: joe@cal.com <j.auyeung419@gmail.com>

joeauyeung committed:
Address feedback

devin-ai-integration[bot] committed:
fix: update teamService integration test to mock new DI factory pattern

Co-Authored-By: joe@cal.com <j.auyeung419@gmail.com>

joeauyeung committed:
Merge branch 'main' into implement-team-billing-repository-factor

hariombalhara commented on packages/features/ee/billing/di/modules/TeamBillingServiceFactory.ts:
Same extension for this file.

hariombalhara commented on packages/features/ee/teams/services/teamService.integration-test.ts:
```diff
@@ -6,16 +6,23 @@ import { MembershipRole } from "@calcom/prisma/enums";
 import { TeamService } from "./teamService";
 
-vi.mock("@calcom/features/ee/billing/teams", () => {
+vi.mock("@calcom/features/ee/billing/di/containers/Billing", () => {
```
The checks are passing

hariombalhara commented on packages/features/ee/billing/repository/teamBillingData/PrismaTeamBillingRepository.ts:
```diff
@@ -0,0 +1,29 @@
+  constructor(private prisma: PrismaClient) {}
+
+  /** Fetch a single team with minimal data needed for billing */
+  async find(teamId: number) {
```
find -> _findUniqueOrThrow_. The name should reflect that it throws.

Similary for findBySubscriptionId below

hariombalhara reviewed (changes requested):
Overall great refactor !! just have very small suggestions

devin-ai-integration[bot] committed:
refactor: remove duplicate imports in credit-service.test.ts

Co-Authored-By: joe@cal.com <j.auyeung419@gmail.com>

joeauyeung committed:
Remove unused index file

joeauyeung committed:
`getBySubscriptionId` to return team or null

joeauyeung committed:
Address feedback

joeauyeung committed:
Merge branch

joeauyeung committed:
Merge fix

alishaz-polymath committed:
Merge branch 'main' into implement-team-billing-repository-factor

joeauyeung committed:
Refactor file names

joeauyeung committed:
Merge with

devin-ai-integration[bot] committed:
fix: correct mockStripe variable name to stripeMock in StripeBillingService.test.ts

Co-Authored-By: joe@cal.com <j.auyeung419@gmail.com>

devin-ai-integration[bot] committed:
refactor: update internal-team-billing.test.ts to use new DI structure with TeamBillingService

- Replace InternalTeamBilling with TeamBillingService
- Use constructor injection with mock dependencies instead of factory pattern
- Remove BillingRepositoryFactory mock and import
- Update all test cases to use mockBillingProviderService, mockTeamBillingDataRepository, and mockBillingRepository
- Simplify saveTeamBilling tests to focus on repository.create calls
- All 11 tests now pass with the new DI structure

Co-Authored-By: joe@cal.com <j.auyeung419@gmail.com>

hariombalhara reviewed (dismissed):
LGTM !! Great refactor @joeauyeung !!

devin-ai-integration[bot] committed:
fix: update createWithPaymentIntent.handler.test.ts to mock DI container's getBillingProviderService

- OrganizationPaymentService now uses getBillingProviderService() from DI container
- Test was mocking @calcom/features/ee/payments/server/stripe directly, which no longer works
- Added mock for @calcom/features/ee/billing/di/containers/Billing module
- Mock returns fake billing provider that delegates to mockSharedStripe
- Preserves all existing test assertions and helpers
- Fixed lint error by prefixing unused lastCreatedSessionId with underscore
- All 11 tests now pass (1 skipped as expected)

Co-Authored-By: joe@cal.com <j.auyeung419@gmail.com>

joeauyeung committed:
Merge branch 'main' into implement-team-billing-repository-factor
